### PR TITLE
[function][Ftr] Add publishFunction for go-function

### PR DESCRIPTION
**Pull request by pantianying**
### Motivation
At present, function only has the function of calculating and changing message content, and has no routing function. This PR is intended to provide such a function. After function calculation, messages can be distributed to different outtopics for different consumers to consume. I think this function is a more practical one.
Explain here the context, and why you're making that change. What is the problem you're trying to solve.

### Modifications

Describe the modifications you've done.

My change is not big. A go function maintains a producer map to replace the previous one. The user can select the producer of the outtopic he wants by setting the value of the outtopic in the context. If the producer does not exist, create one. What I want to discuss here is what we should do if the producer fails to create. as java-function do ,I put this producer map in context

### CODE REVIEW

File: pulsar-function-go/pf/context.go [hunk 1]
```diff
@@ -57,6 +62,35 @@ func (c *FunctionContext) GetOutputTopic() string {
...
+
+func (c *FunctionContext) getProducer(topic string) (pulsar.Producer, error) {
+	if p, ok := c.publishProducers[topic]; ok {
+		return p, nil
+	}
+	properties := getProperties(getDefaultSubscriptionName(
+		c.instanceConf.funcDetails.Tenant,
+		c.instanceConf.funcDetails.Namespace,
+		c.instanceConf.funcDetails.Name), c.instanceConf.instanceID)
+	provider, err := c.instance.client.CreateProducer(pulsar.ProducerOptions{
```
Comment (wolfstudy) [inline pulsar-function-go/pf/context.go]: rename `provider` with `producer`

File: pulsar-function-go/pf/context.go [hunk 2]
```diff
@@ -57,6 +62,35 @@ func (c *FunctionContext) GetOutputTopic() string {
 	return c.instanceConf.funcDetails.GetSink().Topic
 }
 
+func (c *FunctionContext) NewOutputMessage(topicName string) (pulsar.Producer, error) {
```
Comment (wolfstudy) [inline pulsar-function-go/pf/context.go]: Maybe we should return `pulsar.ProducerMessage` instead of `pulsar.Producer`

**Comment by pantianying** [comment]
I've tested it on my cluster machine to match my expectations

**Comment by wolfstudy** [comment]
> I have tested it on my cluster machine and the result is in line with my expectation

Cool, can you add your test case for these changes?

**Comment by wolfstudy** [comment]
@sijie The change LGTM, can you take a look?

File: pulsar-function-go/examples/publishFunc/publish_func.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+package main
+
+import (
+	"context"
+	"github.com/apache/pulsar/pulsar-function-go/pf"
+)
+
+func contextFunc(ctx context.Context) {
+	if fc, ok := pf.FromContext(ctx); ok {
+		fc.NewOutputMessage("your publish topic")
```
Comment (wolfstudy) [inline pulsar-function-go/examples/publishFunc/publish_func.go]: Please remove space or replace space with `-`, because this parameter is a topicName.

**Comment by codelipenghui** [comment]
@pantianying I moved this PR to 2.7.0 If you want to onboard in 2.6.0, feel free to move it back.

**Comment by codelipenghui** [comment]
Looks there is no update on this PR, @wolfstudy could you please help double-check?

**Comment by wolfstudy** [comment]
@pantianying @codelipenghui In https://github.com/apache/pulsar/pull/8327, we have implemented this function, it seems we can close this pull request, welcome @pantianying  to use and verify this function.

**Pull request closed without merging**